source/common/tcp_proxy/tcp_proxy.cc

```diff
@@ -416,23 +421,25 @@ Network::FilterStatus Filter::establishUpstreamConnection() {
 }
 
 void Filter::onClusterDiscoveryCompletion(Upstream::ClusterDiscoveryStatus cluster_status) {
-  // Clear the handle_ before calling establishUpstreamConnection since we may request cluster
-  // again.
+  // Clear the cluster_discovery_handle_ before calling establishUpstreamConnection since we may
+  // request cluster again.
   cluster_discovery_handle_.reset();
   const std::string& cluster_name = route_ ? route_->clusterName() : EMPTY_STRING;
   switch (cluster_status) {
   case Upstream::ClusterDiscoveryStatus::Missing:
-    ENVOY_CONN_LOG(debug, "On on demand cluster {} is missing", read_callbacks_->connection(),
+    ENVOY_CONN_LOG(debug, "On demand cluster {} is missing", read_callbacks_->connection(),
                    cluster_name);
+    config_->stats().on_demand_cluster_missing_.inc();
     break;
   case Upstream::ClusterDiscoveryStatus::Timeout:
-    ENVOY_CONN_LOG(debug, "On on demand cluster {} is not found before time out.",
+    ENVOY_CONN_LOG(debug, "On demand cluster {} was not found before time out.",
                    read_callbacks_->connection(), cluster_name);
+    config_->stats().on_demand_cluster_timeout_.inc();
     break;
   case Upstream::ClusterDiscoveryStatus::Available:
-    if (!downstream_closed_) {
-      establishUpstreamConnection();
-    }
+    // cluster_discovery_handle_ must have cancelled when downstream is closed.
```

Comment:
cluster_discovery_handle_ would have been cancelled if the downstream were closed.